### PR TITLE
[ENH] discrete signature transformer

### LINKED ISSUE #7316: [ENH] discrete signature feature transformer
It would be great to have a simple, no-bells-and-whistles signature feature transformer implemented in `sktime`.

Current implementations suffer either from overengineering (and/or unclear, unstable interfaces), or dependency on abandoned packages relying on lower python versions, see discussion in https://github.com/sktime/sktime/issues/7255

However, the idea behind signatures is quite simple, they are time-ordered generalizations of moments for multivariate time series, I describe it in the following for a bivariate time series.

Let $(x_{1,1}, x_{2,1}) , \dots, (x_{1,T}, x_{2,T})$ be a bivariate time series.

Elements of the "signature" are indexed by strings made up of the literals 1, and 2.
I list the signature elements for the strings up to length 3, the general pattern should follow.

1: the mean of the ${x_{1, i}}$, over all $i$.

2: the mean of the ${x_{2, i}}$, over all $i$.

11: the mean of all products ${x_{1, i}\cdot x_{1, j}}$, over all $i<j$.

12: the mean of all products ${x_{1, i}\cdot x_{2, j}}$, over all $i<j$.

$\ell m$ : the mean of all products ${x_{\ell, i}\cdot x_{m, j}}$, over all $i<j$.

$\ell m n$: the mean of all products ${x_{\ell, i}\cdot x_{m, j}\cdot x_{n, k}}$, over all $i<j<k$.

And so on. Generalisation to longer strings, or more dimensions should be clear.

Due to combinatorial explosion, it typically does not make sense to compute higher than 3, or 2 for high-dimensional time series.

The transformer should be parameterized in terms of an integer `degree`, that is the length of strings to include. Often, the index of the time series is included as a dimension, so for univariate time series, one gets two dimension together with the time index. This should be a boolean option, e.g., `use_index`.

The transformer is "series-to-primitives" type, the supersimple extension template should be sufficient to follow: https://github.com/sktime/sktime/blob/main/extension_templates/transformer_supersimple.py

The `scitypes:transform-output` tag should change to `"Primitives".

### PR DESCRIPTION
**Pull request by AdiTyaPal0710**

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice, this seems to cover all degrees!

Since this is the same idea as `SignatureMoments` transformer, could you try to *replace* or *integrate* your code with it?

Speed is an important point of comparison for the degrees already covered.

File: sktime/transformations/series/signature.py [hunk 1]
```diff
@@ -98,7 +100,7 @@ def _transform(self, X, y=None):
         return pd.DataFrame(signature_matrix, columns=feature_names)
 
     def _compute_signature(self, data):
-        """Compute signature features for a single instance."""
```
Comment (fkiraly) [inline sktime/transformations/series/signature.py:101]: please leave the docstrings as they were.

File: sktime/transformations/series/signature.py [hunk 2]
```diff
@@ -149,10 +151,10 @@ def _compute_mean_product(self, data, indices):
             return data_mean
         else:
             raise NotImplementedError("Degree higher than 3 is not implemented.")
-
+
```
Comment (fkiraly) [inline sktime/transformations/series/signature.py:154]: please ensure to follow the linting guide - there should be no trailing white spaecs
https://www.sktime.net/en/latest/developer_guide/coding_standards.html

File: .gitignore [hunk 3]
```diff
@@ -112,7 +112,7 @@ ENV/
 env.bak/
 venv.bak/
 sktime-dev/
-
+sktime-env/
```
Comment (fkiraly) [inline .gitignore:115]: no need to change the gitignore on the remote

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks a lot!

Could you ensure you replace the existing methods, keeping docstrings etc?

Also it would be great if you could report speed for a few test cases, to see whether the new code is an improvement.

**Comment by AdiTyaPal0710** [comment]
X = pd.DataFrame({
    'x1': [1, 2, 3, 4, 5],
    'x2': [5, 4, 3, 2, 1],
    'x3': [2, 3, 4, 5, 6]
})

for this test case ,the obtained speed was 0.1337 seconds

**Comment by AdiTyaPal0710** [comment]
Alright, I will run some more test cases of different size and compare both the previous and the current code

**Comment by AdiTyaPal0710** [comment]
how about I run on test cases with these size (x,y) :(5,3),(5,4),(10,3),(50,3),(100,5),(500,5) , x = timepoints and y = number of channels ?

**Comment by fkiraly** [comment]
perhaps a wider range of test cases where you vary amount of data, parameters, etc, and measuer secs on your computer

**Comment by AdiTyaPal0710** [comment]
oh alright and from the present test cases the new code is faster compared to the previous ones

**Comment by fkiraly** [comment]
that is great! Could you report a number of test cases with timings, perhaps with some code to reproduce them? That would be great!

**Comment by AdiTyaPal0710** [comment]
Transformation time for previous code 5 timepoints and 3 channels: 0.2166 seconds
Transformation time for previous code 5 timepoints and 4 channels: 0.0137 seconds
Transformation time for previous code 10 timepoints and 3 channels: 0.0131 seconds
Transformation time for previous code 50 timepoints and 3 channels: 0.0144 seconds
Transformation time for previous code 100 timepoints and 5 channels: 0.0268 seconds
Transformation time for previous code 500 timepoints and 5 channels: 0.3946 seconds
-----------------
Transformation time for current code 5 timepoints and 3 channels: 0.0066 seconds
Transformation time for current code 5 timepoints and 4 channels: 0.0064 seconds
Transformation time for current code 10 timepoints and 3 channels: 0.0063 seconds
Transformation time for current code 50 timepoints and 3 channels: 0.0068 seconds
Transformation time for current code 100 timepoints and 5 channels: 0.0090 seconds
Transformation time for current code 500 timepoints and 5 channels: 0.0185 seconds

**Comment by AdiTyaPal0710** [comment]
Transformation time for previous code 5 timepoints and 3 channels: 0.2166 seconds
Transformation time for previous code 5 timepoints and 4 channels: 0.0137 seconds
Transformation time for previous code 10 timepoints and 3 channels: 0.0131 seconds
Transformation time for previous code 50 timepoints and 3 channels: 0.0144 seconds
Transformation time for previous code 100 timepoints and 5 channels: 0.0268 seconds
Transformation time for previous code 500 timepoints and 5 channels: 0.3946 seconds

Transformation time for current code 5 timepoints and 3 channels: 0.0066 seconds
Transformation time for current code 5 timepoints and 4 channels: 0.0064 seconds
Transformation time for current code 10 timepoints and 3 channels: 0.0063 seconds
Transformation time for current code 50 timepoints and 3 channels: 0.0068 seconds
Transformation time for current code 100 timepoints and 5 channels: 0.0090 seconds
Transformation time for current code 500 timepoints and 5 channels: 0.0185 seconds

**Comment by AdiTyaPal0710** [comment]
```python code

sizes = [
    (5, 3), 
    (5, 4), 
    (10, 3), 
    (50, 3), 
    (100, 5),  
    (500, 5),  
]

for n_timepoints, n_channels in sizes:
    X = pd.DataFrame({
        f'x{i+1}': list(range(n_timepoints)) for i in range(n_channels)
    })

    transformer = Sig(degree=2, use_index=True)
    start_time = time.time()
    transformed = transformer.fit_transform(X)
    end_time = time.time()

    print(f"Transformation time for previous code {n_timepoints} timepoints and {n_channels} channels: "
          f"{end_time - start_time:.4f} seconds")

for n_timepoints, n_channels in sizes:
    X = pd.DataFrame({
        f'x{i+1}': list(range(n_timepoints)) for i in range(n_channels)
    })

    transformer = SignatureMoments(degree=2, use_index=True)
    start = time.time()
    transformed = transformer.fit_transform(X)
    end = time.time()

    print(f"Transformation time for current code {n_timepoints} timepoints and {n_channels} channels: "
          f"{end - start:.4f} seconds")


```

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice!

This is really a great improvement!

Some requests:
* it looks like you removed the `normalize_prod` parameter, it would be great if you could add it back - should not be difficult.
* all parameters should be explained in the docstring
* please adhere to the sktime code formatting standards, see here for the guide: https://www.sktime.net/en/stable/developer_guide/coding_standards.html
